### DISCUSSION TITLE
Keep getting 'Saving File Failed' with v 2.16.18.0

### DISCUSSION THREAD

**Discussion by dbowser** [WinMerge/winmerge#1302] [Q&A]
Hello,
WinMerge has worked for me on many systems for many years.
Today I downloaded 2.16.18.0 since I had an older version. 
Now, I cannot save any changes!
I can save in other apps just fine and could with the older version. 
Is there some setting I need to be aware of or something? 
It says it cannot save a backup file, then says it cannot save changes and also cannot copy files. 
I even ran as administrator but that didn't help.

**Comment by sdottaka** [comment]
Does the problem occur with any file?

If you run WinMerge with the `/noprefs` option at the command prompt as follows, does it also fail to save the file?
~~~~
"C:\Program Files\WinMerge\WinMergeU.exe" /noprefs
~~~~

If the file no longer fails to save, then there may be a problem with your configuration. Could you please re-launch WinMerge and attach the configuration information displayed in the Help->Configuration menu item?

**Reply by dbowser** [reply]
I have found that it can save some directories and not others.
Running with ' /noprefs' didn't seem to affect it.
When I check the Properties of the folders that allow it and those that don't both have 'ReadOnly' checked. Clearing it has no effect, it always shows as selected.
When I look at the permissions, Authenticated Users, SYSTEM, and Administrators all have write authority. I am an Administrator.
'Users' do not have write permissions.
I closed the solution in the directory that it couldn't save to in case VS2022 was locking it somehow.

**Reply by sdottaka** [reply]
Could you try running WinMerge as administrator as shown below and see if the file can be saved with WinMerge?

![image](https://user-images.githubusercontent.com/98126/165888735-6316588c-883e-49a0-9817-15ca85d266fc.png)

**Comment by dbowser** [comment]
Hello,
Thank you for trying to help figure this out.
I followed your directions and ran as administrator exactly as you said. 
I made a small edit and tried to save the file.
It still gives an error 'Saving file failed...'.
I opened the file in Notepad++ (not as administrator), made a similar change and it saved without any error.
I tried again in Notepad++ as administrator and it also saved without any error.

**Comment by Gitoffthelawn** [comment]
This is odd, indeed.

Whenever something doesn't work on 1 computer, but others aren't reporting the same issue, there is often something different on that 1 computer.  It's not always the case, but usually is.

With that in mind, here are some questions/thoughts/suggestions:

1. What version of Windows is the system using?
2. Is Windows running in a virtualized environment (within Linux, for example)?
3. Is there a Group Policy (or something equivalent) that could be interfering?
4. Is there security software that could be interfering?
5. Is the processor a common processor (made by Intel or AMD)?
6. Is the target drive encrypted?
7. From where did you download WinMerge?  Did you try downloading it again from GitHub and compare the new download to your original download?
8. It sounds like you have WinMerge set to automatically create backup files.  Have you tried temporarily disabling this option?
9. Does the target partition have plenty of empty space?
10. Have you tried waiting for Mars and Mercury to be aligned in orbit with planet Earth? (okay, just adding a bit of levity on that one) :)

**Comment by sdottaka** [comment]
I recall that file creation was once blocked by security software such as Norton Internet Security.
Is there a record in your security software history that you blocked WinMerge from creating and saving files?

**Reply by hvrmln** [reply]
I have the same problem and know what the cause is but don't know (yet) how to solve it. 
If your configuration is similar to mine, then this may apply to you.

My configuration: Windows 11 (WinMerge says Windows 10), WinMerge 2-16-12-0.
My app is installed as a Node.js server on a Linux-based Raspberry PI (RPI) and as a localhost development version
on the Windows system.
The probem occurs when I have made changes to the development version and then use WinMerge to copy 
them to the RPI version (after stopping the server, of course); this worked fine in earlier WinMerge versions, but 
with WinMerge I can no longer save changed files the backups there; instead it offers to save them on my Windows system and then 
still does not allow me to save the changed RPI files, even though I have made the necessary authorization changes on the RPI. 
On the RPI itself I can use Notepad to make and save the changes manually.

Except for downloading and reinstalling WinMerge, I have checked most of the questions in GitOffthelawn's post (maybe 
he's watching this) and I have tried various configuration options, including manual changes to the configuration file, 
but no luck so far. Unfortunately, WinMerge will not let me tell it not to bother with backups.  
As far as using WinMerge and then saving changed files on the RPI,I don't know how to handle this.

**Reply by Gitoffthelawn** [reply]
@hvrmln wrote:
> ... My app is installed as a Node.js server on a Linux-based Raspberry PI (RPI) and as a localhost development version on the Windows system. ...

Just to ensure I'm understanding correctly, one of your installations of WinMerge is on a RPI and installed as a Node,js server?  Do you then run WinMerge on the RPI and have the output served to other devices?

That's an interesting setup.  Out of curiosity, what is the goal of doing this?  I'm thinking perhaps ease of maintenance?

> ... I have checked most of the questions in GitOffthelawn's post...

I'm glad to hear the list was helpful.  Item 10 is obviously the most important. ;)

@sdottaka Because @hvrmln wrote that it worked in previous WinMerge versions, either something changed in WinMerge or something changed in @hvrmln's setup.  Or, of course, the planets are misaligned again (see item 10 in list above).

**Reply by hvrmln** [reply]
I'm sorry to disappoint you: WinMerge is installed on my Windows system. 
I never considered installing it on the RPI because WinMerge does not 
support Linux.   

My goal of having two 'identical' apps is, indeed, ease of maintenance. 
Both have the same content: information from verious sources
in an easy-to-sort NEDB database (FYI: It's based on JsGrid).

The RPI version is available 24/7 from all my devices (pc, laptop, Chromebook, phone..). 
It is not available on the internet.

The Windows version is something to play with and try out changes and 
additional functions; I sometimes copy the database (RPI to Windows) to
synchronize the two versions and I also sometimes update the RPI version 
with the the Windows changes (Windows to RPI): this is where WinMerge came in handy).

Of course I would prefer to do all of this with WinMerge, but as long the planetes are 
misaligned..... 

Thanks for your feedback and questions.

**Comment by sdottaka** [comment]
Not Rasberry PI, but I installed Samba on Ubuntu 22.04 and compared folders on Ubuntu with WinMerge 2.16.20 and 2.16.12 from Windows and saved the files in those folders without any problems.

**Reply by hvrmln** [reply]
Thanks for your feedback. 
As explained in my post to Gitoffthelawn, I never thought WinMerge would 
work on the RPI. Your Ubuntu approach looks interesting and I may try it though
I am a little worried about the performance.  

I am still puzzled by the fact that I cannot suppress backups on the RPI - I don't really need backups, the Widows version is my backup.
